src/pudl/analysis/record_linkage/classify_plants_ferc1.py

```diff
@@ -0,0 +1,361 @@
+    # which include more than one record from a given year. Warn the user
+    # if we find such cases (which... we do, as of writing)
+    year_dupes = (
+        ferc1_steam_df.groupby(["plant_id_ferc1", "report_year"])["utility_id_ferc1"]
+        .count()
+        .reset_index()
+        .rename(columns={"utility_id_ferc1": "year_dupes"})
+        .query("year_dupes>1")
+    )
+    if len(year_dupes) > 0:
+        for dupe in year_dupes.itertuples():
+            logger.error(
+                f"Found report_year={dupe.report_year} "
+                f"{dupe.year_dupes} times in "
+                f"plant_id_ferc1={dupe.plant_id_ferc1}"
+            )
+    else:
+        logger.info("No duplicate years found in any plant_id_ferc1. Hooray!")
+
+    return ferc1_steam_df
+
+
+@op
+def merge_steam_fuel_dfs(
+    ferc1_steam_df: pd.DataFrame,
+    fuel_fractions: pd.DataFrame,
+) -> pd.DataFrame:
+    """Merge steam plants and fuel dfs to prepare inputs for ferc plant matching."""
+    ffc = list(fuel_fractions.filter(regex=".*_fraction_mmbtu$").columns)
+
+    # Grab fuel consumption proportions for use in assigning plant IDs:
+    return ferc1_steam_df.merge(
+        fuel_fractions[["utility_id_ferc1", "plant_name_ferc1", "report_year"] + ffc],
+        on=["utility_id_ferc1", "plant_name_ferc1", "report_year"],
+        how="left",
+    ).astype({"plant_type": str, "construction_type": str})
+
+
+@graph_asset
+def _out_ferc1__yearly_steam_plants_sched402(
+    plants_steam_ferc1: pd.DataFrame,
+    denorm_fuel_by_plant_ferc1: pd.DataFrame,
+) -> pd.DataFrame:
+    """Assign IDs to the large steam plants."""
+    ###########################################################################
+    # FERC PLANT ID ASSIGNMENT
+    ###########################################################################
+    # Now we need to assign IDs to the large steam plants, since FERC doesn't
+    # do this for us.
+    logger.info("Identifying distinct large FERC plants for ID assignment.")
+
+    input_df = merge_steam_fuel_dfs(plants_steam_ferc1, denorm_fuel_by_plant_ferc1)
+    feature_matrix = ferc_datframe_embedder(input_df)
+    label_df = link_ids_cross_year(input_df, feature_matrix)
+
+    return plants_steam_validate_ids(plants_steam_ferc1, label_df)
+
+
+def revert_filled_in_string_nulls(df: pd.DataFrame) -> pd.DataFrame:
+    """Revert the filled nulls from string columns.
+
+    Many columns that are used for the classification in
+    :func:`plants_steam_assign_plant_ids` have many nulls. The classifier can't handle
+    nulls well, so we filled in nulls with empty strings for string columns. This
+    function replaces empty strings with null values for specific columns that are known
+    to contain empty strings introduced for the classifier.
+    """
+    for col in [
+        "plant_type",
+        "construction_type",
+        "fuel_type_code_pudl",
+        "primary_fuel_by_cost",
+        "primary_fuel_by_mmbtu",
+    ]:
```

Comment:
The revert functions are a bit weird right now. With the new architecture, they never actually need to be used during the matching process, because it doesn't modify the original dataframe. They are still being used in the fuel by plant asset though, so I've kept them for legacy reasons.

Reply:
Got it. That sounds good then.